iothub/device/src/Transport/RetryDelegatingHandler.cs

```diff
@@ -968,7 +968,6 @@ protected override void Dispose(bool disposing)
                 _handleDisconnectCts?.Dispose();
 
                 _handlerSemaphore?.Dispose();
-                _handlerSemaphore = null;
             }
```

Comment:
Line 917 is where the customer got the null reference exception. The order of events seemed to be:

1) They hit a connection loss event that was reported from the mqtt layer.
2) This layer acquired the handlerSemaphore in order to start reopening the mqtt layer.
3) The application layer disposed of this whole client, and set this semaphore to null in doing so
4) When the mqtt reconnection logic finished/failed, it attempted to release the semaphore, but it was null, so a null ref exception was thrown

Reply:
The real customer fix here in my mind is to only dispose of a client that has been closed. That way there is no chance the user would see either the null reference exception (without this change) or the ObjectDisposedException (with this change).

I'd still like to make this change because an ObjectDisposedException makes a lot more sense to throw in the case where a user still decides to dispose a client without closing it first.

Reply:
It's also worth noting that we don't set any other semaphores to null after disposing them, likely for the same reason. See [here](https://github.com/Azure/azure-iot-sdk-csharp/blob/8df9dd76aa6efa87d9085408d221ca61160e1656/iothub/device/src/InternalClient.cs#L1892) and [here](https://github.com/Azure/azure-iot-sdk-csharp/blob/8df9dd76aa6efa87d9085408d221ca61160e1656/iothub/device/src/Transport/Amqp/AmqpUnit.cs#L854)

Reply:
wondering if it makes sense to change behavior (throws ObjectDisposedException) now when you are planning to add real fix (may be soon)?